### PR TITLE
feat: add upgrade functionality for light account to msca

### PR DESCRIPTION
**Pull request by avasisht23**
# Pull Request Checklist

- [ ] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [ ] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [ ] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [ ] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:fix`)
- [ ] Is the base branch you're merging into `development` and not `main`?
- [ ] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

---

### Detailed summary

- Updated `@alchemy/aa-core` dependency to version `1.2.4` in multiple packages.
- Added `encodeUpgradeToAndCall` method to `ISmartContractAccount` interface.
- Added `upgradeAccount` method to `ISmartAccountProvider` interface.
- Added implementation of `encodeUpgradeToAndCall` method to `LightSmartContractAccount` class.
- Added implementation of `upgradeAccount` method to `SmartAccountProvider` class.
- Added tests for upgrading a deployed light account to MSCA.

> The following files were skipped due to too many changes: `packages/accounts/src/msca/utils.ts`, `packages/accounts/src/msca/abis/UpgradeableModularAccount.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/accounts/src/light-account/account.ts
Comment (avasisht23) [inline packages/accounts/src/light-account/account.ts:85]: This is why making the upgrade function generic is difficult. the return type depends on the account type introduced to create a "connected" provider via `.connect`. 

We ran into the same issue with the `createLightAccountAlchemyProvider` factory PR. 

Any ideas on how to genericize without running into this error, which can only be resolved with an unknown cast:
image
Screenshot 2023-11-30 at 8 09 55 PM

File: packages/accounts/src/light-account/e2e-tests/light-account.test.ts
Comment (avasisht23) [inline packages/accounts/src/light-account/e2e-tests/light-account.test.ts:204]: I tested undeployed light account using our gas manager, but now since aa-alchemy optionally depends on aa-accounts, making aa-alchemy a dev dependency of aa-accounts introduces a circular dependency that won't build correctly.

File: packages/accounts/src/light-account/account.ts
Comment (moldy530) [inline packages/accounts/src/light-account/account.ts:28]: nit: this should be configurable and tied to the response of `getDefaultFactoryAddress`

File: packages/accounts/src/msca/utils.ts
Comment (moldy530) [inline packages/accounts/src/msca/utils.ts:89]: we probably also want to install the token receiver plugin once that PR is merged

File: packages/accounts/src/light-account/account.ts
Comment (moldy530) [reply packages/accounts/src/light-account/account.ts:85]: oh I didn't even think about the fact that reconnecting a provider generates a conflict of types... which makes sense because connect is `connect<A>(acct: A) => this & { account: A}` so if you do `connect` when already connected you get `this & { account: A} & {account: B }`

File: packages/accounts/src/msca/types.ts
Comment (avasisht23) [inline packages/accounts/src/msca/types.ts:19]: is `this` necessary? why?

Comment (moldy530) [reply packages/accounts/src/msca/types.ts:19]: yea, so before if you didn't include it then if you previously extended the account you would lose them on the second extend.

If you try it out on the `main` branch right now, you can see that happening in the multi owner builder function. Once you extend with TokenReceiver, the methods for MultiOwner no longer resolve

**Comment by moldy530** [comment]
image

File: packages/core/src/provider/base.ts
Comment (moldy530) [inline packages/core/src/provider/base.ts:772]: I'm gonna clean this up into an actual type and export it

File: packages/accounts/src/light-account/account.ts
Comment (avasisht23) [inline packages/accounts/src/light-account/account.ts:166]: wanna leave a comment to humanize the slot and why this is where we look?

File: packages/accounts/src/light-account/account.ts
Comment (avasisht23) [inline packages/accounts/src/light-account/account.ts:160]: this is probably leftover from me, you can remove it.
```suggestion
```

File: packages/accounts/src/light-account/account.ts
Comment (avasisht23) [inline packages/accounts/src/light-account/account.ts:244]: is there a reason reconnect the provider with the same initCode as the original provider's account? can't we just set the owner to the new owner?

also, do we need to first disconnect at all if we're going to reconnect?

File: packages/accounts/src/msca/utils.ts
Comment (avasisht23) [inline packages/accounts/src/msca/utils.ts:67]: looks like we should remove this?
```suggestion
```

File: packages/accounts/src/light-account/e2e-tests/light-account.test.ts
Comment (avasisht23) [inline packages/accounts/src/light-account/e2e-tests/light-account.test.ts:203]: oh this is cool, but do we think we should just make this an actual field on the returned object and nest the implAddress + initializationData, or leave them as top-line field? do we think devs would ever want to decouple them?

**Review by avasisht23 — COMMENTED** [review]
some open questions, but overall lgtm. looks like you can approve it when you make any necessary changes since it was my og pr haha.

File: packages/accounts/src/light-account/account.ts
Comment (moldy530) [reply packages/accounts/src/light-account/account.ts:166]: ah yea lemme do that

File: packages/accounts/src/light-account/account.ts
Comment (moldy530) [reply packages/accounts/src/light-account/account.ts:244]: yea I don't know if we do need it since we send a UO anyways so it will get deployed with the only account, but just in case I wanted to keep it here

Comment (moldy530) [reply packages/accounts/src/light-account/account.ts:244]: I was disconnecting first since I wanna make some changes to disconnect to clear out the provider methods that connect adds

File: packages/accounts/src/light-account/e2e-tests/light-account.test.ts
Comment (moldy530) [reply packages/accounts/src/light-account/e2e-tests/light-account.test.ts:203]: I decoupled them here because having the connect happen as part of the upgrade call makes typing super hard. so this makes it a bit easier for peeps

File: packages/accounts/src/light-account/account.ts
Comment (moldy530) [inline packages/accounts/src/light-account/account.ts:165]:
```suggestion
      // the slot at which impl addresses are stored by UUPS
      slot: "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
```

**Review by moldy530 — APPROVED** [review]

**Pull request merged**
